src/ImageSharp/Processing/Processors/Quantization/EuclideanPixelMap{TPixel}.cs

```diff
@@ -97,9 +98,111 @@ private int GetClosestColorSlow(TPixel color, ref TPixel paletteRef, out TPixel
+        /// providing enough match accuracy.
+        /// Entry count is currently limited to 1185921 entries (2371842 bytes ~2.26MB).
+        /// </para>
+        /// </remarks>
+        private unsafe struct ColorDistanceCache : IDisposable
+        {
+            private const int IndexBits = 5;
+            private const int IndexAlphaBits = 5;
+            private const int IndexCount = (1 << IndexBits) + 1;
+            private const int IndexAlphaCount = (1 << IndexAlphaBits) + 1;
+            private const int RgbShift = 8 - IndexBits;
+            private const int AlphaShift = 8 - IndexAlphaBits;
+            private const int Entries = IndexCount * IndexCount * IndexCount * IndexAlphaCount;
+            private const int BufferLength = (Entries + 1) >> 1;
+            private MemoryHandle tableHandle;
+            private readonly int[] table;
+            private readonly short* tablePointer;
+
+            private ColorDistanceCache(int bufferLength, int entries)
+            {
+                // We use ArrayPool<int>.Shared for several reasons.
+                // 1. To avoid out of range issues caused by configuring small discontiguous buffers rented via MemoryAllocator
+                // 2. To ensure that the rented buffer is actually pooled.
+                // 3. The .NET runtime already uses this pool so we might already have a pooled array present.
+                this.table = ArrayPool<int>.Shared.Rent(bufferLength);
+                this.tableHandle = this.table.AsMemory().Pin();
+                new Span<short>(this.tableHandle.Pointer, entries).Fill(-1);
+                this.tablePointer = (short*)this.tableHandle.Pointer;
+            }
+
+            public static ColorDistanceCache Create()
+                => new ColorDistanceCache(BufferLength, Entries);
+
+            [MethodImpl(InliningOptions.ShortMethod)]
+            public void Add(Rgba32 rgba, byte index)
+            {
+                int r = rgba.R >> RgbShift;
+                int g = rgba.G >> RgbShift;
+                int b = rgba.B >> RgbShift;
+                int a = rgba.A >> AlphaShift;
+                int idx = GetPaletteIndex(r, g, b, a);
+                this.tablePointer[idx] = index;
+            }
+
+            [MethodImpl(InliningOptions.ShortMethod)]
+            public bool TryGetValue(Rgba32 rgba, out short match)
+            {
+                int r = rgba.R >> RgbShift;
+                int g = rgba.G >> RgbShift;
+                int b = rgba.B >> RgbShift;
+                int a = rgba.A >> AlphaShift;
+                int idx = GetPaletteIndex(r, g, b, a);
+                match = this.tablePointer[idx];
+                return match > -1;
+            }
+
+            [MethodImpl(InliningOptions.ShortMethod)]
+            private static int GetPaletteIndex(int r, int g, int b, int a)
+                => (r << ((IndexBits * 2) + IndexAlphaBits))
```

Comment:
See discussion: https://github.com/SixLabors/ImageSharp/pull/1646#discussion_r643132801

```suggestion
                => (r << ((IndexBits << 1) + IndexAlphaBits))
```

Reply:
This is actually a deoptimization.  JIT can always optimize multiply by powers of 2.  In the case of `x * 2`, it optimizes to either `add` or, even better `lea`, which doesn't set flags.

The comment you linked was referring specifically to signed division, which can't safely be optimized to only a shift because right shift of a negative int rounds toward negative infinity, while a shift of a positive int rounds toward zero.  JIT must therefore emit extra code to force negative values back toward zero after the shift.

[sharplab](https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AXEBDAzgWwB8ABAJgEYBYAKGIGYACMhgYQYG8aHunGBLAHYYGAZT4BzATAAmAET4A3UgApBwhAEoGAXgB8DBAwD0DUgG4uPegzWiJUuYoAcqoQa16Dxhk4vVL3NYArrYAqgL2MvJKyiFumjr6hibmAbwMccLhkY4KLpnuiV4mvmlp1rZikjIAskEANiq2CZ6GAFSmfjzplTl19S7NHkkMHaXU3eWMBdnV0v0qBS0jHakTVtNhEXP9+UNF7T5+AL5AA)

Reply:
Heh, I just noticed JIT will also rewrite `<< 1` as `add` or `lea` as well.  So no harm in writing it either way.  Whatever's more clear will do.

Reply:
> right shift of a negative int rounds toward negative infinity, while a shift of a positive int rounds toward zero

I didn't realize this is the real problem there, two's complement is tricky, thanks!